kani-compiler/src/kani_middle/transform/contracts.rs

```diff
@@ -232,8 +228,297 @@ impl AnyModifiesPass {
+///             kani_register_contract(__kani_recursion_check_modify)
+///         }
+///         kani::internal::REPLACE => {
+///             #[kanitool::is_contract_generated(replace)]
+///             let mut __kani_replace_name_fn = || { /* replace body */ };
+///             kani_register_contract(__kani_replace_name_fn)
+///         }
+///         kani::internal::SIMPLE_CHECK => {
+///             #[kanitool::is_contract_generated(check)]
+///             let mut __kani_check_name_fn = || { /* check body */ };
+///             kani_register_contract(__kani_check_name_fn)
+///         }
+///         _ => { /* original body */ }
+///     }
+/// }
+/// ```
+///
+/// This pass will perform the following operations:
+/// 1. For functions with contract that are not being used for check or replacement:
+///    - Set `kani_contract_mode` to the value ORIGINAL.
+///    - Replace the generated closures body with unreachable.
+/// 2. For functions with contract that are being used:
+///    - Set `kani_contract_mode` to the value corresponding to the expected usage.
+///    - Replace the non-used generated closures body with unreachable.
+/// 3. Replace the body of `kani_register_contract` by `kani::internal::run_contract_fn` to
+///    invoke the closure.
+#[derive(Debug, Default)]
+pub struct FunctionWithContractPass {
+    /// Function that is being checked, if any.
+    check_fn: Option<InternalDefId>,
+    /// Functions that should be stubbed by their contract.
+    replace_fns: HashSet<InternalDefId>,
+    /// Functions annotated with contract attributes will contain contract closures even if they
+    /// are not to be used in this harness.
+    /// In order to avoid bringing unnecessary logic, we clear their body.
+    unused_closures: HashSet<ClosureDef>,
+}
+
+impl TransformPass for FunctionWithContractPass {
+    fn transformation_type() -> TransformationType
+    where
+        Self: Sized,
+    {
+        TransformationType::Stubbing
+    }
+
+    fn is_enabled(&self, _query_db: &QueryDb) -> bool
+    where
+        Self: Sized,
+    {
+        true
+    }
+
+    /// Transform the function body by replacing it with the stub body.
+    fn transform(&mut self, tcx: TyCtxt, body: Body, instance: Instance) -> (bool, Body) {
+        trace!(function=?instance.name(), "FunctionWithContractPass::transform");
+        match instance.ty().kind().rigid().unwrap() {
+            RigidTy::FnDef(def, args) => {
+                if let Some(mode) = self.contract_mode(tcx, *def) {
+                    self.mark_unused(tcx, *def, &body, mode);
+                    let new_body = self.set_mode(tcx, body, mode);
+                    (true, new_body)
+                } else if KaniAttributes::for_instance(tcx, instance).fn_marker()
+                    == Some(Symbol::intern("kani_register_contract"))
+                {
+                    let run = Instance::resolve(find_fn_def(tcx, "KaniRunContract").unwrap(), args)
+                        .unwrap();
```

Comment:
should we use `.expect()` here?

Reply:
And should this be done in the constructor so we only do it once?

Reply:
sure... we can do that